### DISCUSSION TITLE
advanced logging

### DISCUSSION THREAD

**Discussion by Ing-Dom** [thelsing/knx#273] [Ideas]
Hi Thomas,

ich hatte ja eine Überarbeitung des Logging-Konzepts im Stack angekündigt.
Bevor ich jetzt da aber viel Zeit in das Fine-Tuning und dann die Umstellung stecke, würde ich gern kurz deine Meinung dazu haben..

Meine Ziele:
- übersichliches und schnelles Erstellen von Log-Messages (Ein Zeile!)
- Jede Message ist ein Log-Level zugeordnet (Error, Info, Debug, Trace)
- Jeder Message ist ein oder mehrere Log-Bereiche zugeordnet
- keine unnötigen Messages im Binary
- Ausgabe der Messages wie bisher oder alternativ über eine callback-Funktion

Damit kann der Benutzer durch eine Einstellung im Programm auswählen welche Meldungen er sehen will, zB.
- Alle, aber nur Error
- Alle, aber nur Error und Info
- Nur DataLinkLayer, Trace
...

Hier ist ein (noch nicht fertiger) Entwurf:
https://github.com/OpenKNX/knx/tree/draft/advanced_logging

**Comment by thelsing** [comment]
Prinzipiell bin ich dafür. Generell würde ich das gern weniger macrobasiert haben, sondern eher sowas wie:
`class Foo...
static Logger logger = Logger.getClassLogger();
...
logger.debug("foobar");
logger.warning("this is bad");
`
Evtl muss man den Logger pro Klasse in einem statischen Konstruktor mit __FUNC___ intialisieren.

Ich denke vordefinierte Bereiche werden schnell wieder unflexibel und genauer als pro Klasse muss man vermutlich nicht loggen.

In der Logger-Klasse kann man dann mit macro die ganzen nicht benötigten Methodenaufrufe zu NOOPs machen. Die sollte der Compiler dann eigentlich automatisch rausoptimieren, so dass die Strings auch nicht im Binary drin sind.

Evtl. könnte man das sogar nutzen um zur Laufzeit das Logging zu konfigurieren und an eine bestimmte GA zu loggen ;)

**Comment by Ing-Dom** [comment]
Um ehrlich zu sein habe ich da schon einiges an Aufwand spendiert und eine Lösung erarbeitet uind auch getestet die aus meiner Sicht der aktuellen Lösungen in vielerlei Hinsicht überlegen ist.

Es gibt immer viele Wege die nach Rom führen, das ist meiner. Ich werde das nicht nochmal komplett überarbeiten, dazu fehlt mir die Zeit und die Motivation. Über Anpassungen können wir gerne reden, und wenn du konkrete Probleme bei meiner Lösung siehst auch.

Aber nicht über linksrum oder rechtsrum.

Insofern - ist das im Grundsatz so aktzeptabel, oder nicht?

**Comment by thelsing** [comment]
Wenn du über konkreten Code reden willst, mach einfach einen PR, wenn noch nicht fertig ist einen Draft PR dann kann ich mir den Code anschauen und kommentieren.
